File: golang/normal/src/squarprinter.go
package main

import "fmt"

func counter(out chan<- int) {
	for x := 0; x < 10; x++ {
		out <- x
	}
	close(out)
}

func squarer(out chan<- int, in <-chan int) {
	for v := range in {
		out <- v * v
	}
	close(out)
}

func printer(in <-chan int) {
	for v := range in {
		fmt.Println(v)
	}
}

func main() {
	//定义两个int类型的channel
	naturals := make(chan int)
	squares := make(chan int)

	//产生两个Fiber,用go关键字
	go counter(naturals)
	go squarer(squares, naturals)
	//获取计算结果
	printer(squares)
}

File: java/coroutine/demo/src/main/java/com/fhr/coroutine/demo/SquarPrinter.java
package com.fhr.coroutine.demo;

import co.paralleluniverse.fibers.Fiber;
import co.paralleluniverse.fibers.SuspendExecution;
import co.paralleluniverse.strands.channels.Channel;
import co.paralleluniverse.strands.channels.Channels;

/**
 * @author Fan Huaran
 * created on 2018/9/25
 * @description
 */
public class SquarPrinter {

    public static void main(String[] args) throws SuspendExecution, InterruptedException {
        Channel<Integer> naturals = Channels.newChannel(-1);
        Channel<Integer> squares = Channels.newChannel(-1);

        new Fiber<>(() -> counter(naturals)).start();

        new Fiber<>(() -> squarer(naturals, squares)).start();

        printer(squares);
    }

    private static void printer(Channel<Integer> out) throws InterruptedException, SuspendExecution {
        Integer v;
        while ((v = out.receive()) != null) {
            System.out.println(v);
        }
    }

    private static void counter(Channel<Integer> in) throws InterruptedException, SuspendExecution {
        for (int i = 1; i <= 10; i++) {
            in.send(i);
        }
    }

    private static void squarer(Channel<Integer> in, Channel<Integer> out) throws InterruptedException, SuspendExecution {
        Integer v;
        while ((v = in.receive()) != null) {
            out.send(v * v);
        }
    }

}
